src/value.rs

```diff
@@ -0,0 +1,188 @@
+    pub const fn u2(value: u8) -> Result<Self, Error> {
+        match value {
+            0..=3 => Ok(Self::U2(value)),
+            _ => Err(Error::IntegerOutOfBounds(UIntType::U2)),
+        }
+    }
+
+    /// Create a value of type [`u4`].
+    ///
+    /// ## Errors
+    ///
+    /// Value is greater than 15.
+    pub const fn u4(value: u8) -> Result<Self, Error> {
+        match value {
+            0..=15 => Ok(Self::U4(value)),
+            _ => Err(Error::IntegerOutOfBounds(UIntType::U4)),
+        }
+    }
+
+    /// Create an integer from a `decimal` string and `ty`pe.
+    pub fn parse_decimal(decimal: &UnsignedDecimal, ty: UIntType) -> Result<Self, Error> {
+        let s = decimal.as_inner();
+        match ty {
+            UIntType::U1 => s.parse::<u8>().map_err(Error::from).and_then(Self::u1),
+            UIntType::U2 => s.parse::<u8>().map_err(Error::from).and_then(Self::u2),
+            UIntType::U4 => s.parse::<u8>().map_err(Error::from).and_then(Self::u4),
+            UIntType::U8 => s.parse::<u8>().map_err(Error::from).map(Self::U8),
+            UIntType::U16 => s.parse::<u16>().map_err(Error::from).map(Self::U16),
+            UIntType::U32 => s.parse::<u32>().map_err(Error::from).map(Self::U32),
+            UIntType::U64 => s.parse::<u64>().map_err(Error::from).map(Self::U64),
+            UIntType::U128 => s.parse::<u128>().map_err(Error::from).map(Self::U128),
+            UIntType::U256 => s.parse::<U256>().map_err(Error::from).map(Self::U256),
+        }
+    }
+}
+
+impl From<u8> for UIntValue {
+    fn from(value: u8) -> Self {
+        Self::U8(value)
+    }
+}
+
+impl From<u16> for UIntValue {
+    fn from(value: u16) -> Self {
+        Self::U16(value)
+    }
+}
+
+impl From<u32> for UIntValue {
+    fn from(value: u32) -> Self {
+        Self::U32(value)
+    }
+}
+
+impl From<u64> for UIntValue {
+    fn from(value: u64) -> Self {
+        Self::U64(value)
+    }
+}
+
+impl From<u128> for UIntValue {
+    fn from(value: u128) -> Self {
+        Self::U128(value)
+    }
+}
+
+impl From<U256> for UIntValue {
+    fn from(value: U256) -> Self {
+        Self::U256(value)
+    }
+}
+
+impl TryFrom<&[u8]> for UIntValue {
+    type Error = &'static str;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        match value.len() {
+            1 => Ok(Self::U8(value[0])),
+            2 => Ok(Self::U16(u16::from_be_bytes(value.try_into().unwrap()))),
+            4 => Ok(Self::U32(u32::from_be_bytes(value.try_into().unwrap()))),
+            8 => Ok(Self::U64(u64::from_be_bytes(value.try_into().unwrap()))),
+            16 => Ok(Self::U128(u128::from_be_bytes(value.try_into().unwrap()))),
+            32 => Ok(Self::U256(U256::from_byte_array(value.try_into().unwrap()))),
+            _ => Err("Too many bytes"),
+        }
+    }
+}
+
+impl From<Bits> for UIntValue {
+    fn from(value: Bits) -> Self {
+        match value {
+            Bits::U1(byte) => Self::u1(byte).expect("Always <= 1"),
```

Comment:
In a786308a9bb84f7dc2e0b82dd3aec23d72bd9f81:

FWIW this isn't actually guaranteed because `Bits` is a public enum and users can construct it wrong if they want.

Reply:
Mmh. We might want to force users through a constructor that checks the invariants by making the enum private...